src/java_bytecode/java_bytecode_parser.cpp

```diff
@@ -1509,3 +1636,54 @@ void java_bytecode_parsert::parse_local_variable_type_table(methodt &method)
       "Entry in LocalVariableTypeTable must be present in LVT");
   }
 }
+
+/// Read method handle pointed to from constant pool entry at index, return type
+/// of method handle and name if lambda function is found.
+/// \param entry: the constant pool entry of the methodhandle_info structure
+/// \param[out]: the method_handle type of the methodhandle_structure, either
+/// for a recognized bootstrap method, for a lambda function or unknown
+void java_bytecode_parsert::parse_methodhandle(
+  const pool_entryt &entry,
+  lambda_method_handlet &handle)
+{
+  INVARIANT(
+    entry.tag == CONSTANT_MethodHandle,
+    "constant pool entry must be a MethodHandle");
+  const auto &ref_entry = pool_entry(entry.ref2);
+  INVARIANT(
+    (entry.ref1 > 0 && entry.ref1 < 10),
+    "reference kind of Methodhandle must be in the range of 1 to 9");
+
+  const auto &class_entry = pool_entry(ref_entry.ref1);
```

Comment:
This bit is kind of complex to follow, suggest mirroring the spec to first create an enum for the different values of the `reference_kind`:

```
REF_getField
REF_getStatic
REF_putField
REF_putStatic
REF_invokeVirtual
REF_invokeStatic
REF_invokeSpecial
REF_newInvokeSpecial
REF_invokeInterface
```
Then switch on these in the same way, breaking into  `rFieldref_info`

I'm pausing here as it has got hard to follow - but the special casing of dealing with which factories we actually support should be abstracted far away from the parse tree IMO (in fact potentially not even part of this PR?)